### PR TITLE
[Feature](agg_state) support agg_state combinator on nereids

### PR DESCRIPTION
**Pull request by BiteTheDDDDt**
## Proposed changes

upport agg_state combinator on nereids

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by BiteTheDDDDt** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggStateFunctionBuilder.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+    private BoundFunction buildState(String nestedName, List<? extends Object> arguments) {
+        return nestedBuilder.build(nestedName, arguments);
+    }
+
+    private BoundFunction buildMergeOrUnion(String nestedName, List<? extends Object> arguments) {
+        if (arguments.size() != 1 || !(arguments.get(0) instanceof Expression)
+                || !((Expression) arguments.get(0)).getDataType().isAggStateType()) {
+            String argString = arguments.stream().map(arg -> {
+                if (arg == null) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggStateFunctionBuilder.java:74]: why arg could be null?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggStateFunctionBuilder.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                    return "null";
+                } else if (arg instanceof Expression) {
+                    return ((Expression) arg).toSql();
+                } else {
+                    return arg.toString();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggStateFunctionBuilder.java:79]: why arg is not Expression?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggStateFunctionBuilder.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            throw new IllegalStateException("Can not build AggState nested function: '" + nestedName + "', expression: "
+                    + nestedName + argString);
+        }
+
+        Expression arg = (Expression) arguments.get(0);
+        AggStateType type = (AggStateType) arg.getDataType();
+
+        List<Expression> nestedArgumens = type.getSubTypes().stream().map(t -> {
+            return new SlotReference("mocked", t);
+        }).collect(Collectors.toList());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggStateFunctionBuilder.java:91]:
```suggestion
        List<Expression> nestedArgumens = type.getSubTypes().stream()
                .map(t -> new SlotReference("mocked", t))
                .collect(ImmutableList.toImmutableList());
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggStateFunctionBuilder.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        return combinatorSuffix + "(" + nestedBuilder.toString() + ")";
+    }
+
+    public static boolean isAggStateCombinator(String name) {
+        return name.toLowerCase().endsWith(COMBINATOR_LINKER + STATE)
+                || name.toLowerCase().endsWith(COMBINATOR_LINKER + MERGE)
+                || name.toLowerCase().endsWith(COMBINATOR_LINKER + UNION);
+    }
+
+    public static String getNestedName(String name) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggStateFunctionBuilder.java:119]: should check name contains COMBINATOR_LINKER first, and throw exception if not contains it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggStateFunctionBuilder.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        return name.substring(0, name.lastIndexOf(COMBINATOR_LINKER));
+    }
+
+    public static String getCombinatorSuffix(String name) {
+        return name.substring(name.lastIndexOf(AggStateFunctionBuilder.COMBINATOR_LINKER) + 1);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/AggStateFunctionBuilder.java:124]: should check name contains COMBINATOR_LINKER first, and throw exception if not contains it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/ExpressionTranslator.java [hunk 6]
```diff
@@ -430,10 +442,44 @@ public Expr visitIsNull(IsNull isNull, PlanTranslatorContext context) {
...
+
+    @Override
+    public Expr visitMergeCombinator(MergeCombinator combinator, PlanTranslatorContext context) {
+        List<Expr> arguments = combinator.getArguments().stream().map(arg -> arg.accept(this, context))
+                .collect(Collectors.toList());
+        return Function.convertToMergeCombinator(new FunctionCallExpr(
+                translateAggregateFunction(combinator.getNestedFunction()), new FunctionParams(false, arguments)));
+    }
+
+    private org.apache.doris.catalog.AggregateFunction translateAggregateFunction(AggregateFunction function) {
```
[the remark this answers is not in the document]
Comment (BiteTheDDDDt) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/ExpressionTranslator.java]: This logic can be used for both `translateAggregateFunction`/`visitStateCombinator`/`visitMergeCombinator`

**Comment by BiteTheDDDDt** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Pull request merged**